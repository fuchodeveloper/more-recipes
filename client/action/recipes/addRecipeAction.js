import { batchActions } from 'redux-batched-actions';
import axios from 'axios';
import alertify from 'alertify.js';
import { ADD_RECIPE, ADD_RECIPE_ERROR } from '../types';
import { setFetching, unsetFetching } from '../fetching';

/**
 * @description create a recipe
 *
 * @export addRecipeActionCreator
 *
 * @param {Object} recipe
 *
 * @returns {Object} recipe
 */
const addRecipeActionCreator = recipe => ({
  type: ADD_RECIPE,
  recipe
});

/**
 * @description create a recipe error
 *
 * @export addRecipeActionError
 *
 * @param {Object} error
 *
 * @returns {Object} error
 */
const addRecipeActionError = error => ({
  type: ADD_RECIPE_ERROR,
  error
});

const addRecipeAction = recipe => (dispatch) => {
  dispatch(setFetching());
  return axios.post('/api/v1/recipes', recipe)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      dispatch(batchActions([
        dispatch(addRecipeActionCreator(response.data.recipe)),
        unsetFetching()
      ]));
    })
    .catch((error) => {
      dispatch([
        dispatch(addRecipeActionError(error.response.data.error)),
        unsetFetching()
      ]);
    });
};

export default addRecipeAction;
