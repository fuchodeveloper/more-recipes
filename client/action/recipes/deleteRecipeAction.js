import axios from 'axios';
import alertify from 'alertify.js';
import { DELETE_RECEIPE, DELETE_RECEIPE_FAIL } from '../types';
import networkError from '../networkError';

export const deleteRecipeActionError = error => ({
  type: DELETE_RECEIPE_FAIL,
  error
});

export const deleteRecipeActionCreator = recipes => ({
  type: DELETE_RECEIPE,
  recipes
});

const deleteRecipeAction = recipeId => dispatch =>
  axios.delete(`/api/v1/recipes/${recipeId}`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      dispatch(deleteRecipeActionCreator(response.data.recipes));
    })
    .catch((error) => {
      if (!error.response) {
        return networkError(error);
      }
      dispatch(deleteRecipeActionError(error.response.data));
    });

export default deleteRecipeAction;
