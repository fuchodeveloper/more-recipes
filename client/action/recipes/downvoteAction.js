import axios from 'axios';
import alertify from 'alertify.js';
import { GET_DOWNVOTE } from '../types';

/**
 * Recipe doenvote action creator
 * @param {recipe} recipe
 * @returns {recipe} recipe
 */
export const downvoteRecipeActionCreator = recipe => ({
  type: GET_DOWNVOTE,
  recipe
});

/**
 * Authenticated user can downvote a recipe
 *
 * @export func
 * @param {integer} param
 * @returns {object} object
 */
const downvoteRecipe = param =>
  dispatch => axios.post(`/api/v1/recipes/${param}/downvote`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      return dispatch(downvoteRecipeActionCreator(response.data.recipe));
    })
    .catch((error) => {
      alertify.logPosition('bottom right');
      alertify.error(error.message);
    });


export default downvoteRecipe;
