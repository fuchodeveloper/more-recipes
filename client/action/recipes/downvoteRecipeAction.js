import axios from 'axios';
import alertify from 'alertify.js';
import { DOWNVOTE_RECIPE, DOWNVOTE_RECIPE_ERROR } from '../types';
import networkError from '../networkError';

/**
 * @description Recipe downvote action creator
 *
 * @param {recipe} recipe recipe parameter
 *
 * @returns {recipe} recipe downvoted
 */
export const downvoteRecipeActionCreator = recipe => ({
  type: DOWNVOTE_RECIPE,
  recipe
});
/**
 * @description Recipe downvote error
 *
 * @param {Object} error error parameter
 *
 * @returns {Object} error returns the error object
 *
 */
const downvoteRecipeActionError = error => ({
  type: DOWNVOTE_RECIPE_ERROR,
  error
});

/**
 * @description Authenticated user can downvote a recipe
 *
 * @export downvoteRecipe
 *
 * @param {Number} id
 *
 * @returns {object} downvoted recipe object
 */
const downvoteRecipeAction = id =>
  dispatch => axios.post(`/api/v1/recipes/${id}/downvote`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.recipe.message);
      return dispatch(downvoteRecipeActionCreator(response.data.recipe));
    })
    .catch((error) => {
      if (!error.response) {
        return networkError(error);
      }
      alertify.delay(1000);
      alertify.logPosition('bottom right');
      alertify.error('Please login to downvote recipe');
      dispatch(downvoteRecipeActionError(error.response.data));
    });


export default downvoteRecipeAction;
