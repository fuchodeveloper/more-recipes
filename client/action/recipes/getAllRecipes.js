import axios from 'axios';
import { batchActions } from 'redux-batched-actions';
import { GET_ALL_RECIPES, GET_PAGE_DETAILS } from '../types';
import { setFetching, unsetFetching } from '../fetching';

export const receiveRecipes = recipes => ({
  type: GET_ALL_RECIPES,
  recipes
});

export const recipePageDetails = pageCount => ({
  type: GET_PAGE_DETAILS,
  pageCount
});

const getAllRecipes = page => (dispatch) => {
  dispatch(setFetching());
  axios.get(`/api/v1/recipes?page=${page}`)
    .then((response) => {
      dispatch(batchActions([
        receiveRecipes(response.data.recipes),
        recipePageDetails(response.data.pageCount),
        unsetFetching(),
      ]));
    })
    .catch((error) => {
      // dispatch(batchActions([

      // ]));
      console.log(error.message);
    });
};

export default getAllRecipes;
