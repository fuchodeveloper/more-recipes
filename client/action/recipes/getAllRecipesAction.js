import axios from 'axios';
import { batchActions } from 'redux-batched-actions';
import
{ GET_ALL_RECIPES, GET_PAGE_DETAILS, GET_ALL_RECIPES_ERROR } from '../types';
import { setFetching, unsetFetching } from '../fetching';

/**
 * @description get all recipes action creator
 *
 * @param {Object} recipes recipes object parameter
 *
 * @returns {Object} recipes returns all recipes object
 */
const receiveRecipes = recipes => ({
  type: GET_ALL_RECIPES,
  recipes
});

/**
 * @description get all recipes error action
 *
 * @param {Object} error recipes error object
 *
 * @returns {Object} error returns get all recipes error object
 */
const receiveRecipesError = error => ({
  type: GET_ALL_RECIPES_ERROR,
  error
});

/**
 * @description handles page detail
 *
 * @param {Number} pageCount recipe page count parameter
 *
 * @returns {Number} returns recipes page Count
 */
const recipePageCount = pageCount => ({
  type: GET_PAGE_DETAILS,
  pageCount
});

const getAllRecipesAction = page => (dispatch) => {
  dispatch(setFetching());
  return axios.get(`/api/v1/recipes?page=${page}`)
    .then((response) => {
      dispatch(batchActions([
        dispatch(receiveRecipes(response.data.recipes)),
        recipePageCount(response.data.pageCount),
        unsetFetching(),
      ]));
    })
    .catch((serverError) => {
      dispatch(receiveRecipesError(serverError.recipes.data.error));
    });
};

export default getAllRecipesAction;
