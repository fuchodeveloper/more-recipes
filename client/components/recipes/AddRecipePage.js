import React from 'react';
import Header from '../../components/navigation/Header';
import Footer from '../../components/navigation/Footer';
import AddRecipeForm from './AddRecipeForm';

class AddRecipePage extends React.Component {
  render() {
    return (
      <div>
        {/* Header component for navigation */}
        <Header />

        <div className="container margin-top-70">
        <div>
            <h1 className="text-center p-4 center-hero-text">Add Recipe</h1>
        </div>

        {/* Recipe avatar */}
        <div className="col-md-6 mx-auto p-3 add_recipe-card text-center">
            <img src="./img/no-image.svg" width="350px" id="recipe-image-avatar" height="auto" className="img-fluid" alt=""/>
        </div>

        {/* Form to add recipe ingredients and quantity */}
        <div className="col-md-6 mx-auto p-3">
          
          <AddRecipeForm />

        </div>
    </div>

    <div className="clearfix m-5"/>

    <Footer />

  </div>
    );
  }
}

export default AddRecipePage;
