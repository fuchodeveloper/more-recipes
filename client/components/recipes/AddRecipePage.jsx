/* eslint-disable no-undef, max-len */
import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import addRecipeAction from '../../action/recipes/addRecipeAction';
import { placeholderImage } from '../../assets/img/helperImage';
import validateRecipe from '../../validations/validateRecipe';

/**
 * @description container to add a new recipe
 *
 * @class AddRecipePage
 *
 * @extends {React.Component}
 */
export class AddRecipePage extends React.Component {
  /**
   * @description Creates an instance of AddRecipePage.
   *
   * @param {Object} props constructor props
   *
   * @memberof AddRecipePage
   */
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      image: placeholderImage,
      ingredients: '',
      direction: '',
      errors: {
        name: '',
        ingredients: '',
        direction: '',
        image: ''
      },
      redirect: false
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.uploadWidget = this.uploadWidget.bind(this);
  }
  /**
 * @description component lifecycle method
 *
 * @param {Object} nextProps componentWillReceiveProps nextProps object
 *
 * @memberof AddRecipePage
 *
 * @returns {undefined} sets state of errors
 */
  componentWillReceiveProps(nextProps) {
    const { name, ingredients, direction } =
    (nextProps.errors.serverError) ? nextProps.errors.serverError : {};
    this.setState({
      errors: {
        name, ingredients, direction
      }
    });
  }

  /**
 * @description Function to handle change event
 *
 * @param {Object} event onChange event object
 *
 * @memberof AddRecipePage
 *
 * @returns {undefined} sets state on change
 */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
 * @description function to handle form submission
 *
 * @param {Object} event onSubmit event object
 *
 * @memberof AddRecipePage
 *
 * @returns {undefined} calls addRecipeAction on submit
 */
  onSubmit(event) {
    event.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {} });
      this.props.addRecipeAction(this.state)
        .then(() => {
          this.setState({ redirect: true });
        });
    }
  }

  /**
   * @description Handle form validation
   *
   * @returns {Object} isValid returns boolean
   *
   * @memberof AddRecipePage
   */
  isValid() {
    const { errors, isValid } = validateRecipe(this.state);

    if (!isValid) {
      this.setState({ errors });
    }

    return isValid;
  }

  /**
 * @description function to handle image uploads
 *
 * @param {Object} event image event object
 *
 * @memberof AddRecipePage
 *
 * @returns {undefined} new image upload
 */
  uploadWidget(event) {
    event.preventDefault();
    const scope = this;

    cloudinary.openUploadWidget(
      {
        cloud_name: 'fuchodeveloper',
        upload_preset: 'wvxnziq0',
        tags: ['recipe']
      },
      (error, result) => {
        if (!error) {
          scope.setState({ image: result[0].secure_url });
        }
      }
    );
  }

  /**
 * @description renders component to DOM
 *
 * @memberof AddRecipePage
 *
 *  @returns {JSX} JSX representation of component
 */
  render() {
    if (this.state.redirect) {
      return <Redirect to="/" />;
    }

    return (
      <div>
        <div className="container margin-top-70">
          <div>
            <h1 className="text-center p-4 center-hero-text">Add Recipe</h1>
          </div>


          {/* Form to add recipe ingredients and quantity */}
          <div className="col-md-6 mx-auto p-3">
            {/* Recipe avatar */}
            <div className="col-md-6 mx-auto p-3 add_recipe-card text-center">

              <img
                src={this.state.image === ''
                ? placeholderImage : this.state.image}
                width="350px"
                id="recipe-image-avatar"
                height="auto"
                className="img-fluid"
                alt={this.state.name}
              />
            </div>

            <form
              encType="multipart/form-data"
              onSubmit={this.onSubmit}
              id="add-recipe-form"
            >
              <br />
              <div className="upload text-center">
                <button
                  name="image"
                  id="image-button"
                  onClick={this.uploadWidget}
                  className="upload-button btn-primary-color text-white"
                >
                      Add Image
                </button>
              </div>

              <h3 className="text-center p-4 center-hero-text">
              Add Ingredients
              </h3>

              <div className="form-row" id="recipe-quantity">
                <div className="col-sm-12">
                  <div>
                    <p className="text-muted">Recipe Name</p>
                    <input
                      placeholder="Enter recipe name"
                      type="text"
                      name="name"
                      id="name"
                      className="form-control"
                      value={this.state.name}
                      onChange={this.onChange}
                      required
                    />

                    { this.state.errors.name &&
                      <div className="text-danger">
                        {this.state.errors.name}
                      </div>
                    }

                  </div>
                </div>
                <div className="m-1" />

                <div className="col-sm-12">
                  <div>
                    <br />
                    <p className="text-muted">Recipe Ingredients</p>
                    <textarea
                      placeholder="e.g Maggi, Knorr, pepper"
                      cols="30"
                      rows="5"
                      name="ingredients"
                      id="ingredients"
                      className="form-control"
                      value={this.state.ingredients}
                      onChange={this.onChange}
                      required
                    />
                    { this.state.errors.ingredients &&
                      <div className="text-danger">
                        {this.state.errors.ingredients}
                      </div>
                    }
                  </div>
                </div>
              </div>

              <div className="clearfix" />
              <div className="form-row">
                <div className="col-md-12 mb-3">
                  <label htmlFor="direction" />
                  <p className="text-muted">Recipe Direction</p>
                  <textarea
                    className="form-control"
                    placeholder="Give a detailed direction for this recipe:"
                    name="direction"
                    id="recipe-direction"
                    cols="30"
                    rows="10"
                    value={this.state.direction}
                    onChange={this.onChange}
                    required
                  />
                  { this.state.errors.direction &&
                  <div className="text-danger">
                    {this.state.errors.direction}
                  </div>
                    }
                </div>
              </div>

              <div className="float-right p-1">
                <button
                  type="submit"
                  className="btn btn-primary btn-primary-color"
                  id="recipe-submit"
                >Submit
                </button>
              </div>

            </form>
          </div>
        </div>

        <div className="clearfix m-5" />
      </div>
    );
  }
}

AddRecipePage.propTypes = {
  addRecipeAction: PropTypes.func.isRequired
};

AddRecipePage.contextTypes = {
  router: PropTypes.object
};

const mapStateToProps = state => ({
  errors: state.recipesReducer
});

const mapDispatchToProps = dispatch => ({
  addRecipeAction: recipe => dispatch(addRecipeAction(recipe))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddRecipePage);
