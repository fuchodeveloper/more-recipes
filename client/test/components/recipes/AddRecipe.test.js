import React from 'react';
import { shallow } from 'enzyme';
import expect from 'expect';
import sinon from 'sinon';
import { AddRecipePage } from '../../../components/recipes/AddRecipePage';
import mockData from '../../__mocks__/mockData';

let props;

const setup = () => {
  props = {
    cloudinary: {
      openUploadWidget: jest.fn(() => Promise.resolve())
    },
    addRecipeAction: jest.fn(() => Promise.resolve())
  };
  return shallow(<AddRecipePage {...props} />);
};

describe('AddRecipePage component', () => {
  describe('AddRecipePage component snapshot', () => {
    it('should render with right amount of elements', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });
  it(
    'should test that the AddRecipePage component rendered successfully',
    () => {
      const wrapper = setup();
      expect(wrapper.find('div').length).toBe(17);
      expect(wrapper.find('form').length).toBe(1);
      expect(wrapper.find('img').length).toBe(1);
      expect(wrapper.find('input').length).toBe(1);
      expect(wrapper.find('textarea').length).toBe(2);
    }
  );

  it('should set the recipe name when input value changes', () => {
    const event = {
      target: { name: 'name', value: '' }
    };
    const wrapper = setup();
    const recipeName = wrapper.find('#name');

    event.target.value = 'jollof rice';
    recipeName.simulate('change', event);

    expect(wrapper.instance().state.name).toBe('jollof rice');
  });

  it('should set the recipe ingredients when input value changes', () => {
    const event = {
      target: { name: 'ingredients', value: '' }
    };
    const wrapper = setup();
    const recipeName = wrapper.find('#ingredients');

    event.target.value = 'rice and maggi';
    recipeName.simulate('change', event);

    expect(wrapper.instance().state.ingredients).toBe('rice and maggi');
  });

  it('should not add recipe if state is empty', () => {
    const event = {
      preventDefault: jest.fn()
    };
    const wrapper = setup();
    const form = wrapper.find('#add-recipe-form');

    form.simulate('submit', event);
  });

  it('should set the recipe image on click and call uploadWidget', () => {
    const event = {
      preventDefault: jest.fn()
    };

    const wrapper = setup();
    const recipeImage = wrapper.find('#image-button');

    recipeImage.simulate('click', event);
    wrapper.setState({
      image: 'https://res.cloudinary.com/fuchodeveloper/image/upload/'
      + 'v1516760699/noodles_c6ltkq.jpg'
    });
    expect(wrapper.instance().uploadWidget(event)).toMatchSnapshot();
  });
  
  it('should add new recipe if state is set currently', () => {
    const event = {
      preventDefault: jest.fn()
    };
    const { addRecipe } = mockData;
    const wrapper = setup();
    const form = wrapper.find('#add-recipe-form');
    wrapper.setState(addRecipe);

    expect(wrapper.instance().state.name).toEqual(addRecipe.name);
    form.simulate('submit', event);
  });

  it('component should receive props', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const spy = sinon.spy(
      AddRecipePage.prototype,
      'componentWillReceiveProps'
    );
    const nextProps = {
      errors: {
        serverError: {
          name: 'Recipe name is required',
          ingredients: 'Recipe ingredients is required',
          direction: 'Recipe ingredients is required'
        }
      },
    };
    shallow(<AddRecipePage {...props} componentWillReceiveProps={spy} />);
    action.componentWillReceiveProps(nextProps);
    expect(action.state.errors.name)
      .toBe(nextProps.errors.serverError.name);
    expect(action.state.errors.ingredients)
      .toBe(nextProps.errors.serverError.ingredients);
    expect(action.state.errors.direction)
      .toBe(nextProps.errors.serverError.direction);
  });
});

