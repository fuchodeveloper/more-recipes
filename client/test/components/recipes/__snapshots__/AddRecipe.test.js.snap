// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecipePage component AddRecipePage component snapshot should render with right amount of elements 1`] = `undefined`;

exports[`AddRecipePage component should set the recipe image on click and call uploadWidget 1`] = `undefined`;
